{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genomic Data for Variant Pathogenicity\n",
    "This notebook reads the vcf file containing ClinVar data and outputs a vcf file that contains the right information to run ANNOVAR and, eventually reach the table templated format provide in FH-EARLY for the genomic data.\n",
    "\n",
    "### To download the ClinVar data:\n",
    "Go to https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/ (Our version is clinvar_20260208.vcf.gz and also clinvar_20260208.vcf.gz.tbi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import packages\n",
    "import pandas as pd\n",
    "from cyvcf2 import VCF # https://github.com/brentp/cyvcf2/tree/main"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sources for choices\n",
    "1. Genome Interval (ClinVar): consulted https://www-ncbi-nlm-nih-gov.tudelft.idm.oclc.org/clinvar/?term=LDLR%5Bgene%5D to select the range between the first and the last LDLR variant in chromosome 19.\n",
    "2. Variables ANNOVAR (following MetaRNN paper - additional file):\n",
    "    * chromosome = var.CHROM\n",
    "    * position = var.POS\n",
    "    * reference allele = var.REF\n",
    "    * alternative allele = var.ALT\n",
    "    * [missing] reference aa of the protein [gloria]\n",
    "    * [missing] alternative aa of the protein [gloria]\n",
    "    * label (TP or TN) (Clinical Significance by CLinVar)= var.INFO.get('CLNSIG','')\n",
    "3. Additional variables used in FH-EARLY:\n",
    "    * [uncertain] function = for now var.INFO.get('MC','') (eg. synonymous_variant)\n",
    "    * gene = var.INFO.get('GENEINFO', '')\n",
    "    * variant type = var.INFO.get('CLNVC','')\n",
    "    * [uncertain] change = for now included in var.INFO.get('CLNHGVS','') (look for >)\n",
    "    * ONIM = included in var.INFO.CLNDISDB (look for ONIM)\n",
    "4. Additional variables needed to select samples\n",
    "    * id = var.ID\n",
    "    * star review = var.INFO.get('CLNREVSTAT','')\n",
    "5. Additional variables we need to get output from 1000 genomes, gnomAD exome, and dbSNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ALT', 'CHROM', 'FILTER', 'FILTERS', 'FORMAT', 'ID', 'INFO', 'POS', 'QUAL', 'REF', '__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', 'aaf', 'call_rate', 'end', 'format', 'genotype', 'genotypes', 'gt_alt_depths', 'gt_alt_freqs', 'gt_bases', 'gt_depths', 'gt_phases', 'gt_phred_ll_het', 'gt_phred_ll_homalt', 'gt_phred_ll_homref', 'gt_quals', 'gt_ref_depths', 'gt_types', 'is_deletion', 'is_indel', 'is_mnp', 'is_snp', 'is_sv', 'is_transition', 'nucl_diversity', 'num_called', 'num_het', 'num_hom_alt', 'num_hom_ref', 'num_unknown', 'ploidy', 'relatedness', 'set_format', 'set_pos', 'start', 'var_subtype', 'var_type']\n"
     ]
    }
   ],
   "source": [
    "vcf_og = VCF('clinvar_20260208.vcf.gz')\n",
    "print(dir(next(iter(vcf_og))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>review</th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>clinsig</th>\n",
       "      <th>function</th>\n",
       "      <th>type</th>\n",
       "      <th>change</th>\n",
       "      <th>onim</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>694275</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>19</td>\n",
       "      <td>11087729</td>\n",
       "      <td>ACCACGCCCGGCTAATTTTTTGTATTTTTTTTTAGTAGAGGTGGGG...</td>\n",
       "      <td>A</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td></td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NC_000019.10:g.11087732_11090710del</td>\n",
       "      <td>Human_Phenotype_Ontology:HP:0003124,Human_Phen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>430740</td>\n",
       "      <td>reviewed_by_expert_panel</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>19</td>\n",
       "      <td>11089263</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>SO:0001619|non-coding_transcript_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>NC_000019.10:g.11089263C&gt;G</td>\n",
       "      <td>MONDO:MONDO:0005439,MedGen:C0020445,OMIM:PS143...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>250925</td>\n",
       "      <td>criteria_provided,_multiple_submitters,_no_con...</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>19</td>\n",
       "      <td>11089281</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Benign/Likely_benign</td>\n",
       "      <td>SO:0001619|non-coding_transcript_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>NC_000019.10:g.11089281G&gt;T</td>\n",
       "      <td>MONDO:MONDO:0005439,MedGen:C0020445,OMIM:PS143...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4069787</td>\n",
       "      <td>no_assertion_criteria_provided</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>19</td>\n",
       "      <td>11089283</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>SO:0001619|non-coding_transcript_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>NC_000019.10:g.11089283C&gt;T</td>\n",
       "      <td>MONDO:MONDO:0005439,MedGen:C0020445,OMIM:PS143890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3628882</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>19</td>\n",
       "      <td>11089309</td>\n",
       "      <td>CAG</td>\n",
       "      <td>C</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>SO:0001619|non-coding_transcript_variant</td>\n",
       "      <td>Microsatellite</td>\n",
       "      <td>NC_000019.10:g.11089310AG[1]</td>\n",
       "      <td>MONDO:MONDO:0005439,MedGen:C0020445,OMIM:PS143890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4319</th>\n",
       "      <td>328126</td>\n",
       "      <td>criteria_provided,_conflicting_classifications</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>19</td>\n",
       "      <td>11133635</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Conflicting_classifications_of_pathogenicity</td>\n",
       "      <td>SO:0001624|3_prime_UTR_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>NC_000019.10:g.11133635C&gt;G</td>\n",
       "      <td>MedGen:C3661900|MONDO:MONDO:0007750,MedGen:C07...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4320</th>\n",
       "      <td>890682</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>19</td>\n",
       "      <td>11133666</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>SO:0001624|3_prime_UTR_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>NC_000019.10:g.11133666C&gt;G</td>\n",
       "      <td>MONDO:MONDO:0007750,MedGen:C0745103,OMIM:14389...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4321</th>\n",
       "      <td>328127</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>19</td>\n",
       "      <td>11133681</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>SO:0001624|3_prime_UTR_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>NC_000019.10:g.11133681C&gt;T</td>\n",
       "      <td>MONDO:MONDO:0007750,MedGen:C0745103,OMIM:14389...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4322</th>\n",
       "      <td>328128</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>19</td>\n",
       "      <td>11133682</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>SO:0001624|3_prime_UTR_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>NC_000019.10:g.11133682G&gt;T</td>\n",
       "      <td>MONDO:MONDO:0007750,MedGen:C0745103,OMIM:14389...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4323</th>\n",
       "      <td>328129</td>\n",
       "      <td>criteria_provided,_single_submitter</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>19</td>\n",
       "      <td>11133700</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>SO:0001624|3_prime_UTR_variant</td>\n",
       "      <td>single_nucleotide_variant</td>\n",
       "      <td>NC_000019.10:g.11133700G&gt;T</td>\n",
       "      <td>MONDO:MONDO:0007750,MedGen:C0745103,OMIM:14389...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4324 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id                                             review  \\\n",
       "0      694275                criteria_provided,_single_submitter   \n",
       "1      430740                           reviewed_by_expert_panel   \n",
       "2      250925  criteria_provided,_multiple_submitters,_no_con...   \n",
       "3     4069787                     no_assertion_criteria_provided   \n",
       "4     3628882                criteria_provided,_single_submitter   \n",
       "...       ...                                                ...   \n",
       "4319   328126     criteria_provided,_conflicting_classifications   \n",
       "4320   890682                criteria_provided,_single_submitter   \n",
       "4321   328127                criteria_provided,_single_submitter   \n",
       "4322   328128                criteria_provided,_single_submitter   \n",
       "4323   328129                criteria_provided,_single_submitter   \n",
       "\n",
       "                              gene chrom       pos  \\\n",
       "0     LDLR:3949|LDLR-AS1:115271120    19  11087729   \n",
       "1     LDLR:3949|LDLR-AS1:115271120    19  11089263   \n",
       "2     LDLR:3949|LDLR-AS1:115271120    19  11089281   \n",
       "3     LDLR:3949|LDLR-AS1:115271120    19  11089283   \n",
       "4     LDLR:3949|LDLR-AS1:115271120    19  11089309   \n",
       "...                            ...   ...       ...   \n",
       "4319                     LDLR:3949    19  11133635   \n",
       "4320                     LDLR:3949    19  11133666   \n",
       "4321                     LDLR:3949    19  11133681   \n",
       "4322                     LDLR:3949    19  11133682   \n",
       "4323                     LDLR:3949    19  11133700   \n",
       "\n",
       "                                                    ref alt  \\\n",
       "0     ACCACGCCCGGCTAATTTTTTGTATTTTTTTTTAGTAGAGGTGGGG...   A   \n",
       "1                                                     C   G   \n",
       "2                                                     G   T   \n",
       "3                                                     C   T   \n",
       "4                                                   CAG   C   \n",
       "...                                                 ...  ..   \n",
       "4319                                                  C   G   \n",
       "4320                                                  C   G   \n",
       "4321                                                  C   T   \n",
       "4322                                                  G   T   \n",
       "4323                                                  G   T   \n",
       "\n",
       "                                           clinsig  \\\n",
       "0                                       Pathogenic   \n",
       "1                           Uncertain_significance   \n",
       "2                             Benign/Likely_benign   \n",
       "3                           Uncertain_significance   \n",
       "4                           Uncertain_significance   \n",
       "...                                            ...   \n",
       "4319  Conflicting_classifications_of_pathogenicity   \n",
       "4320                        Uncertain_significance   \n",
       "4321                        Uncertain_significance   \n",
       "4322                        Uncertain_significance   \n",
       "4323                        Uncertain_significance   \n",
       "\n",
       "                                      function                       type  \\\n",
       "0                                                                Deletion   \n",
       "1     SO:0001619|non-coding_transcript_variant  single_nucleotide_variant   \n",
       "2     SO:0001619|non-coding_transcript_variant  single_nucleotide_variant   \n",
       "3     SO:0001619|non-coding_transcript_variant  single_nucleotide_variant   \n",
       "4     SO:0001619|non-coding_transcript_variant             Microsatellite   \n",
       "...                                        ...                        ...   \n",
       "4319            SO:0001624|3_prime_UTR_variant  single_nucleotide_variant   \n",
       "4320            SO:0001624|3_prime_UTR_variant  single_nucleotide_variant   \n",
       "4321            SO:0001624|3_prime_UTR_variant  single_nucleotide_variant   \n",
       "4322            SO:0001624|3_prime_UTR_variant  single_nucleotide_variant   \n",
       "4323            SO:0001624|3_prime_UTR_variant  single_nucleotide_variant   \n",
       "\n",
       "                                   change  \\\n",
       "0     NC_000019.10:g.11087732_11090710del   \n",
       "1              NC_000019.10:g.11089263C>G   \n",
       "2              NC_000019.10:g.11089281G>T   \n",
       "3              NC_000019.10:g.11089283C>T   \n",
       "4            NC_000019.10:g.11089310AG[1]   \n",
       "...                                   ...   \n",
       "4319           NC_000019.10:g.11133635C>G   \n",
       "4320           NC_000019.10:g.11133666C>G   \n",
       "4321           NC_000019.10:g.11133681C>T   \n",
       "4322           NC_000019.10:g.11133682G>T   \n",
       "4323           NC_000019.10:g.11133700G>T   \n",
       "\n",
       "                                                   onim  \n",
       "0     Human_Phenotype_Ontology:HP:0003124,Human_Phen...  \n",
       "1     MONDO:MONDO:0005439,MedGen:C0020445,OMIM:PS143...  \n",
       "2     MONDO:MONDO:0005439,MedGen:C0020445,OMIM:PS143...  \n",
       "3     MONDO:MONDO:0005439,MedGen:C0020445,OMIM:PS143890  \n",
       "4     MONDO:MONDO:0005439,MedGen:C0020445,OMIM:PS143890  \n",
       "...                                                 ...  \n",
       "4319  MedGen:C3661900|MONDO:MONDO:0007750,MedGen:C07...  \n",
       "4320  MONDO:MONDO:0007750,MedGen:C0745103,OMIM:14389...  \n",
       "4321  MONDO:MONDO:0007750,MedGen:C0745103,OMIM:14389...  \n",
       "4322  MONDO:MONDO:0007750,MedGen:C0745103,OMIM:14389...  \n",
       "4323  MONDO:MONDO:0007750,MedGen:C0745103,OMIM:14389...  \n",
       "\n",
       "[4324 rows x 12 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows = []\n",
    "for var in vcf_og('19:11087732-11133700'):\n",
    "    # TODO: select necessary information\n",
    "    rows.append({\n",
    "        # identifiers\n",
    "        'id': var.ID,\n",
    "        'review': var.INFO.get('CLNREVSTAT',''),\n",
    "        'gene': var.INFO.get('GENEINFO', ''),\n",
    "        'chrom': var.CHROM,\n",
    "        # useful\n",
    "        'pos': var.POS,  # 1-based\n",
    "        'ref': var.REF,\n",
    "        'alt': ','.join(var.ALT), # this because it is a lisy\n",
    "        # missing: 'ref_aa': \n",
    "        # missing: 'alt_aa':\n",
    "        'clinsig': var.INFO.get('CLNSIG', ''),\n",
    "        'function': var.INFO.get('MC',''),\n",
    "        'type': var.INFO.get('CLNVC',''),\n",
    "        'change': var.INFO.get('CLNHGVS',''),\n",
    "        'onim': var.INFO.get('CLNDISDB','')\n",
    "    })\n",
    "\n",
    "df = pd.DataFrame(rows)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* only samples encoding LDLR variants\n",
    "* id is unique\n",
    "* pos is almost unique, there are some variants in the same position"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# information needed for ANNOVAR\n",
    "1. chromosome\n",
    "2. start position (pos)\n",
    "3. end position - TODO\n",
    "4. ref nucleotide (ref)\n",
    "5. observed nucleotide (alt)\n",
    "\n",
    "## TO CHECK: I think ANNOVAR is only for missense (1 nucleotide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "catt",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
